src/common/entity/compute_state_display.ts

```diff
@@ -28,38 +28,73 @@ export const computeStateDisplay = (
 
   const domain = computeStateDomain(stateObj);
 
+  // Below logic for `input_datetime` works only if the frontend's (browser's) timezone is the same as core's.
+  // If frontend's timezone is different from core's, `input_datetime`s' state display computation will be incorrect.
+  // Needs further fixes for frontend to get core's timezone offset, in order to make computation correct all the time.
   if (domain === "input_datetime") {
-    let date: Date;
-    if (!stateObj.attributes.has_time) {
+    if (state) {
+      // If trying to display an explicit state, need to parse the explict state to `Date` then format.
+      // Attributes aren't available, we have to use `state`.
+      try {
+        if (!stateObj.attributes.has_time) {
+          // Only has date.
+          const dateObj = new Date(state);
+          // When only date is passed to `Date` constructor, it uses UTC regardless of frontend's timezone,
+          // so we need to add the frontend's timezone offset.
+          // Other usages of `Date` constructor have both date and time in the string, frontend's timezone is used w/o problem.
+          const timezoneOffset = new Date().getTimezoneOffset() * 60 * 1000;
+          const adjustedDateObj = new Date(dateObj.getTime() + timezoneOffset);
```

Comment:
<img width="1905" alt="snipaste_20210612_150853" src="https://user-images.githubusercontent.com/7324708/121788245-71826700-cb91-11eb-81ce-b577155f0e2d.png">

Reply:
Safari does not accept this format for the Date constructor at all, it will return `Invalid Date`